SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialLoadBalanceEnforcer.cpp

```diff
@@ -1,271 +1,214 @@
+		PopulateDataStore(EntityId);
+		RefreshAcl(EntityId);
 	}
+}
 
-	const SpatialGDK::AuthorityIntent* AuthorityIntentComponent =
-		StaticComponentView->GetComponentData<SpatialGDK::AuthorityIntent>(EntityId);
-	const PhysicalWorkerName* OwningWorkerId =
-		VirtualWorkerTranslator->GetPhysicalWorkerForVirtualWorker(AuthorityIntentComponent->VirtualWorkerId);
+void SpatialLoadBalanceEnforcer::RefreshAcl(const Worker_EntityId EntityId)
+{
+	const AuthorityIntent& AuthorityIntentComponent = DataStore[EntityId].Intent;
 
-	check(OwningWorkerId != nullptr);
-	if (OwningWorkerId == nullptr)
-	{
-		UE_LOG(LogSpatialLoadBalanceEnforcer, Error,
-			   TEXT("Couldn't find mapped worker for entity %lld. This shouldn't happen! Virtual worker ID: %d"), EntityId,
-			   AuthorityIntentComponent->VirtualWorkerId);
-		return;
-	}
+	check(VirtualWorkerTranslator != nullptr);
+	const PhysicalWorkerName* DestinationWorkerId =
+		VirtualWorkerTranslator->GetPhysicalWorkerForVirtualWorker(AuthorityIntentComponent.VirtualWorkerId);
 
-	if (*OwningWorkerId == WorkerId && StaticComponentView->HasAuthority(EntityId, SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID))
+	check(DestinationWorkerId != nullptr);
+	if (DestinationWorkerId == nullptr)
 	{
-		UE_LOG(LogSpatialLoadBalanceEnforcer, Verbose,
-			   TEXT("No need to queue newly authoritative entity because this worker is already authoritative. Entity: %lld. Worker: %s."),
-			   EntityId, *WorkerId);
 		return;
 	}
 
-	if (AclAssignmentRequestIsQueued(EntityId))
-	{
-		UE_LOG(LogSpatialLoadBalanceEnforcer, Verbose,
-			   TEXT("Avoiding queueing a duplicate ACL assignment request. Entity: %lld. Worker: %s."), EntityId, *WorkerId);
-		return;
-	}
-
-	QueueAclAssignmentRequest(EntityId);
-}
-
-bool SpatialLoadBalanceEnforcer::AclAssignmentRequestIsQueued(const Worker_EntityId EntityId) const
-{
-	return AclWriteAuthAssignmentRequests.Contains(EntityId);
+	UpdateSender(ConstructAclUpdate(EntityId, DestinationWorkerId));
 }
 
-TArray<SpatialLoadBalanceEnforcer::AclWriteAuthorityRequest> SpatialLoadBalanceEnforcer::ProcessQueuedAclAssignmentRequests()
+EntityComponentUpdate SpatialLoadBalanceEnforcer::ConstructAclUpdate(const Worker_EntityId EntityId,
+																	 const PhysicalWorkerName* DestinationWorkerId)
 {
-	TArray<SpatialLoadBalanceEnforcer::AclWriteAuthorityRequest> PendingRequests;
+	LBComponents& Components = DataStore[EntityId];
+	EntityAcl& Acl = Components.Acl;
+	const ComponentPresence& ComponentPresenceComponent = Components.Presence;
+	const NetOwningClientWorker& NetOwningClientWorkerComponent = Components.OwningClientWorker;
 
-	TArray<Worker_EntityId> CompletedRequests;
-	CompletedRequests.Reserve(AclWriteAuthAssignmentRequests.Num());
+	TArray<Worker_ComponentId> ComponentIds;
+	Acl.ComponentWriteAcl.GetKeys(ComponentIds);
 
-	for (Worker_EntityId EntityId : AclWriteAuthAssignmentRequests)
+	// Ensure that every component ID in ComponentPresence is set in the write ACL.
```

Comment:
When would we expect these two lists to not be consistent with each other?

Reply:
dynamic component additions from another worker

- actor-auth worker: wants to add component. Updates component presence component with the new component.
- ACL-auth worker: sees this component presence update. Ensures the current actor-auth worker is now delegated this new component in the ACL. (Previously it wasn't)
- actor-auth worker gains auth over the component it previously specified, and can now add it.